pkg/codec/byte_string_modifier.go

```diff
@@ -0,0 +1,466 @@
+package codec
+
+import (
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"reflect"
+
+	"golang.org/x/crypto/sha3"
+
+	"github.com/mr-tron/base58"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/types"
+)
+
+type AddressLength int
+type ChecksumType string
+type EncodingType string
+
+const (
+	// EVM
+	Byte20Address AddressLength = 20
+	EIP55         ChecksumType  = "eip55"
+	HexEncoding   EncodingType  = "hex"
+	// Solana
+	Byte32Address  AddressLength = 32
+	Base58Encoding EncodingType  = "base58"
+	// General
+	NoneChecksum ChecksumType = "none"
+)
+
+type AddressChecksum func([]byte) []byte
+
+// EIP55AddressChecksum Applies EIP55 checksum logic to the address
+// assuming input 'a' is already in hex form without the "0x" prefix
+func EIP55AddressChecksum(a []byte) []byte {
```

Comment:
How is this chain-agnostic? The gold standard is that we are able to integration new chains, only by authoring a LOOP Plugin, and without having to update any other modules, like chainlink/ or this one. This code looks _universal_ to me, meaning that it knows about every chain. That doesn't meet our requirements.

Reply:
Thought the requirement was not importing libraries like `go-ethereum` directly.

Removing the `Checksum` part would make this development comply with the standards? Length and encoding seem something general that could be reused by different blockchains

Reply:
I think we need to zoom out. What problem are we trying to solve? This repository is imported by core, all products, and all chains. Core and all products are chain-agnostic, and so cannot use this. Where would it be called from?

Reply:
We want to write a codec modifier that parse addresses into strings, the issue is that all of the previous modifiers aren't doing anything chain specific and live in common. The modifiers get created based on chain agnostic codec configs [here](https://github.com/smartcontractkit/chainlink-common/blob/e1435d915916aef78e985938f9cc98fe501507f2/pkg/codec/config.go#L41), which directly initialise a chain agnostic modifier struct [here](https://github.com/smartcontractkit/chainlink-common/blob/e1435d915916aef78e985938f9cc98fe501507f2/pkg/codec/config.go#L184). 

One solution that I think could work(not completely sure), would be to inject a chain specific implementation of a modifier during Contract Reader init, just leave some kind of an empty interface in the modifier in common and set it when Contract Reader is initialised in the relayer.

Anothe solution is to remove change this modifier to support all of these formats(I just put together a random list):
- EVM is        20 bytes Base16 (Hexadecimal)
- Solana is     32 bytes Base58
- Starknet is  32 bytes Base16 (Hexadecimal)
-  Cosmos is  20 bytes Bech32
- Aptos is       32 bytes Base16 (Hexadecimal)
- Polkadot is  32 bytes SSBase58
- Cardano is   32 bytes Bech32
- Tezos is       20 bytes Base58Check
- Algorand is  32 bytes Base32
- Avalanche is 20 bytes Bech32
- Tron is           20 bytes Base58Check
- Near is          32 bytes Base58

I don't know cryptography well enough to determine if this approach makes sense and if chains have a reason to introduce new encodings

Reply:
> One solution that I think could work(not completely sure), would be to have that config call into a modifier interface and then to inject a chain specific implementation during Contract Reader init, just leave some kind of an empty interface in the modifier in common and set it when Contract Reader is initialised in the relayer.

This sounds right. I don't see any need to parameterize anything here. The relayers know what format to expect. I think this kind of extensible API was inevitable, because the codec runs on the server side.